[ENH] Automated peak finding in Spectra

noahcabanas opened the pull request:
I added in a new functionality to the spectra menu to label peaks automatically based off of the scipy find peaks function. In tangent with this is the functionality to add individual lines.

markotoplak commented on orangecontrib/spectroscopy/utils/Peak_line.py:
```diff
@@ -0,0 +1,569 @@
+"""
+
+
+class PeakLine(GraphicsObject):
```
Thanks for the PR. This file mostly seems to be a copy of the Infinite Line from pyqtgraph. This will be very hard to main. Instead, the infinite line should be subclassed, because the small modification the subclass will do are likely to be sufficient.

If subclassing is not possible, all changes to the pyqtgraph code need to clearly marked with a very visible comment.

markotoplak commented on orangecontrib/spectroscopy/utils/Peak_line.py:
```diff
@@ -0,0 +1,569 @@
+
+class PeakLine(GraphicsObject):
+    """
+    **Bases:** :class:`GraphicsObject <pyqtgraph.GraphicsObject>`
```
For subclassing or encapsulation of InfiniteLine see Orange.widgets.evaluate.owrocanalysis.InfiniteLine (subclassing with small changes) or orangecontrib.spectroscopy.widgets.gui.MovableVLine (for encapsulation with friendlier signals). Is there any particular reason why you did not use MovableVLine (or its slight modification) for your PR?

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
+                    peak_locations.append(peak[j])
+                    # find any peaks which aren't duplicates(None) or nearby mislabels
+            for k, val in enumerate(peak_locations):
+                Label_line = pl.PeakLine()
```
Creating new peaks should be a function, because now, the same code pattern repeats 3 times. Perhaps peak_apply could the function if it was given proper arguments.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
+                                                           maxHeight]), prominence=prominence)
+            peaks = x_axis[peaks]
+            for i, val in enumerate(peaks):
+                Label_line = pl.PeakLine()
```
Same, this should call a function (a large chunk of code is repeating)

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
+
+    def peak_apply_auto(self, prominence, minHeight, maxHeight, line_overlap):
+        x_axis = getx(self.data)
+        data = np.array(self.data)
```
f you need spectral values, there are in data.X. Data can also contain other meta or class data.

Otherwise, self.data_x already contains the sorted x_axis, and data.X[:, self.data_xsind] would give you data with already sorted wavenumbers. I think you find_peaks function expects ordered inputs, while a data table, after transformations in Orange widgets  became unordered wavenumber-wise. Here it is important to use the ordered variants.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1062,101 @@ def grid_apply(self):
+        data = np.array(self.data)
+        peak = []
+        if np.shape(data) != (len(data),):
+            for i, val in enumerate(data):
```
Did you perhaps try your code on data with 100000 spectra? Please do.

I am just guessing, but grouping of same peaks will need to be written much more efficiently. Perhaps you could sort the list and then go through it only once. You algorithm complexity is O(n**2) where n is number of all peaks in the whole data set, which will not fly. It needs to be at most O(nlogn), achievable with sorting + inspection of the sorted list. There are probably better ways to do it as I thought of now, though.

Also, if there is a huge data set, looking at its subset of spectra for peaks will probably be sufficient.

markotoplak commented:
Thank you, @noahcabanas, this is a much needed feature.

I looked through the code and left some comments that could help you make the code more robust, faster, and maintainable.

Writing code that is short and does not repeat is really important from the maintenance perspective. For any code we add to the repo, we increase the maintenance burden (which is, I think, for any project the majority of work involved after some time). So we really try to keep all additions as clean as possible.

markotoplak commented:
Also, maybe I have missed it, but I do not recall any mechanism for saving the peak markings. Ideally, if you save a workflow with marked peaks and then open it on a different computer, the same peaks should be shown. But this is likely the least important of my comments, please tackle this one last.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1095,15 +1095,16 @@ def peak_apply(self, position):
-        x_axis = getx(self.data)
-        data = np.array(self.data)
+        x_axis = self.data_x
+        data = self.data
```
I think the data here is still problematic. See my previous comment. We have to insure, that the code also works in the presence of meta variable and class attributes, this we need to use just X. So it could be `data.X`. But there is an additional problem: wavenumbers might not be ordered. Therefore I suggested `data = data.X[:, self.data_xsind]` in the previous comment. That should be robust.

markotoplak commented on doc/widgets/labeling.md:
```diff
@@ -0,0 +1,21 @@
+# Peak Labeling
+
+The peak labeling menu enables allows for individual or automatic peak labeling.
```
Reduce wordiness and try to make writing clearer. Perhaps use Grammarly for help?

markotoplak commented on doc/widgets/labeling.md:
```diff
@@ -0,0 +1,21 @@
+to adjust what is labeled as a peak even when plotting multiple spectra with
+overlapping peak positions. This process is limited by the clarity of data.
+As such, while it can preform on noisy datasets, it will yield clear results
+after data has been filtered. Added peaks can be remvoed by double right clicking. 
```
Double-clicking is not something that can be easily discovered. Possible solutions, ideally both:
1. Right click on the line and the delete option in the submenu. 
2. An option to select the line by clicking, this selection should have a visual indicator and them pressing delete or backspace for removal.

markotoplak commented on orangecontrib/spectroscopy/tests/test_owspectra.py:
```diff
@@ -146,6 +146,18 @@ def test_select_line(self):
             sa = out.transform(Domain([out.domain[ANNOTATED_DATA_FEATURE_NAME]]))
             np.testing.assert_equal(sa.X, 1)
 
+    def test_Label_peaks(self):
```
Add peak label tests that tests on ugly data. Whenever we write code we have to make as few assumptions about the data as possible. If a user merged two datasets, wavenumbers would not be continuous and there are going to be many unknown values. So, test also with data that:
1) Contains unknowns, lots of them, also some rows and columns that are completely unknown.
2) Has a mixed order of wavenumbers.

The code will surely have problems with such data, because I saw amax used which does not account for NaNs.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+            self.span = (mn, mx)
+            self.update()
+
+    def _computeBoundingRect(self):
```
Also, this function is mostly copied, what is the difference? Is the difference needed at all? If needed, but predictable, maybe the `super()` function can be called and then the output could be adapted.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
         return QWidget.focusInEvent(self, *e)
+
+
+class VerticalPeakLine(pg.InfiniteLine):
```
This subclass still redefines too much (copies from pg.InfiniteLine). The first superfluous line is `self.moving = False` and it is just the first.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+            self.label = VerticalPeakLineLabel(self,
+                                               text=str(self.getXPos()), position=(1))
+
+    def setSpan(self, mn, mx):
```
This is exact the same function as in the superclass and should be removed.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+
+        return self._bounds
+
+    def boundingRect(self):
```
The same as in superclass and therefore needs to be removed.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+        p.drawLine(Point(left, 0), Point(right, 0))
+
+
+    def mouseDragEvent(self, ev):
```
Should not be modified. The only added thing is `updateLabel` could be connected to the sigDragged signal in `__init__`.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+            self.sigDragged.emit(self)
+            self.sigPositionChangeFinished.emit(self)
+
+    def mouseDoubleClickEvent(self, ev):
```
Double clicking to remove is something we should not use. We need to stick with established UI paradigms.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+        self.setText(self.format.format(value=value))
+        self.updatePosition()
+
+    def mouseDragEvent(self, ev):
```
With a quick check I do not see any difference. What is different? If nothing, this function can be removed.

markotoplak commented on orangecontrib/spectroscopy/widgets/gui.py:
```diff
@@ -474,3 +477,152 @@ def set_default(self, v):
+        super().__init__(line)
+        self.line = line
+
+    def valueChanged(self):
```
No changes compared to the super class, should be removed.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
 
+    def find_peak_variables(self):
+        if self.maximum_point is None:
+            self.maximum_point = np.amax(self.data.X[:, self.data_xsind])
```
amax is problematic for data with unknown values (nans)

markotoplak replied:
Also, ordering is not needed here.

markotoplak replied:
Also, does it really have to be computed from the whole data set or just for the data shown? I'd say only on the data shown because this is much faster. And then these can be recomputed when the curves that are shown are resampled.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -668,6 +665,13 @@ class CurvePlot(QWidget, OWComponent, SelectionGroupMixin):
+    peak_min = Setting(None, schema_only=True)
+    peak_max = Setting(None, schema_only=True)
+    peak_locations = Setting([], schema_only=True)
+    minimum_point = Setting(None, schema_only=True)
```
minimum_point, maximum_points should not be a setting, because they are always defined by data and never changed by the UI.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -694,12 +698,12 @@ def __init__(self, parent: OWWidget, select=SELECTNONE):
         self.subset = None  # current subset input, an array of indices
         self.subset_indices = None  # boolean index array with indices in self.data
-
+        self.start_point = None  # automatic position for adding single lines in peak_apply
```
`start_point` should not be a module attribute, as it can be quickly computed when needed (is totally data dependent).

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -694,12 +698,12 @@ def __init__(self, parent: OWWidget, select=SELECTNONE):
         self.subset_indices = None  # boolean index array with indices in self.data
-
+        self.start_point = None  # automatic position for adding single lines in peak_apply
+        self.peak_state = 0  # Value for determining what our peak clear should do
```
Why does `peak_state` do? What do its values mean? It is not clear from the code why it is needed, I'd guess it is really not. Peak reappearance should be handled differently.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -809,6 +811,46 @@ def __init__(self, parent: OWWidget, select=SELECTNONE):
+        self.reset_labels = QPushButton('Reset graph', self)
+        self.single_peak.clicked.connect(self.find_peak_variables)
+        self.single_peak.clicked.connect(self.peak_apply)
+        self.reset_labels.clicked.connect(self.change_state)
```
`change_state` should be removed. Peak updating needs to be handled differently.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+        if self.minimum_point is None:
+            self.minimum_point = np.amin(self.data.X[:, self.data_xsind])
+        if self.start_point is None:
+            self.start_point = round(np.median(getx(self.data)), 2)
```
Do not round here. Only round strings at the output. Who knows what is the user's range (perhaps it can be 0.0000001 to 0.0000300).

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+
+    def peak_apply(self, position):
+        if self.viewtype == INDIVIDUAL:
+            Label_line = VerticalPeakLine()
```
`Label_line` is not a PEP-8 name.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+            data = self.data.X[:, self.data_xsind]
+            peak = []
+            if np.shape(data) != (len(data),):
+                for i, val in enumerate(data):
```
We should never go though the whole dataset, which can be VERY large, just through the curves shown.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+                for k, val in enumerate(peak_locations):
+                    self.peak_apply(position=val)
+            else:
+                # covers singular spectra data instances
```
Do not duplicate code for single and multiple spectra. Code that works for multiple spectra also needs to work for a single spectrum. Maybe this remained from a previous version.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+                        peak.append(var)
+                peak_locations = []
+                sorted_peaks = np.sort(peak)
+                for i, val in enumerate(sorted_peaks):
```
peak_locations needs to be a set so that `in` predicate operates in constant amortized time.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+                        # if its greater than than nearby values
+                        peak_locations.append(val)
+                    elif not peak_locations:
+                        peak_locations.append(val)
```
This is not critical and can always be improved later, just as a note: when peaks are now close, the first one is selected. It would be better if we selected the "middle" one.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1128,7 +1259,7 @@ def current_selection():
 
         # redraw whole selection if it increased or decreased over the threshold
         if len(old_sel_ci) <= MAX_THICK_SELECTED < len(new_sel_ci) or \
-           len(old_sel_ci) > MAX_THICK_SELECTED >= len(new_sel_ci):
```
Do not modify the code you did not need to edit. And if you do, ALWAYS do it in a separate commit.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1074,6 +1197,14 @@ def clear_graph(self):
         for m in self.markings:
             self.plot.addItem(m, ignoreBounds=True)
 
+        if self.peak_state != 1:
```
`peak_state` needs to be removed and drawing solved differently.

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+            Label_line = VerticalPeakLine()
+            Label_line.setMovable(True)
+            Label_line.setPen(pg.mkPen(color=QColor(Qt.black), width=2, style=Qt.DotLine))
+            Label_line.setSpan(mn=self.minimum_point, mx=self.maximum_point)
```
Why is this setSpan needed? Why not draw Infinite lines also out of the data set?

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1009,6 +1067,71 @@ def grid_apply(self):
+            if position:
+                Label_line.setPos(position)
+                Label_line.label.setText(str(round(position, 3)))
+            else:
```
What if a user zoomed-in before adding peak lines? Then the line should appear in the middle of the zoomed-in view. Now it appears in the middle of the data.

markotoplak commented:
@noahcabanas, thanks for your PR. Sorry for not getting to it sooner.

The code was substantially improved compared to the previous versions, thanks. There is still some work to d (I wrote individual comments). To summarize:
- The code is currently not robust to unknown values or strange cases (like no data in the widget, then it crashes) and is also not tested as such. Perhaps the easiest way to test is to call the automatic peak labeler in a function similar to the existing `test_average` that tests on obnoxious data.
- Any code performing computation should only work on the spectra shown (that is a random subset of 1000 spectra) because who knows how big is the user's data set.
- We should have as few settings and object properties as possible. If something can be inferred from the data, it should not be a setting. Or object property, except in rare cases where we aim to save computation time (but that needs a very good reason for it). We try to avoid object properties too, because they need careful updating.
- The subclasses of InfiniteLine and the Label should be simplified further.
- The `peak_state` property and the drawing tricks should be removed. These need to be solved differently.

Please tell me if you are still prepared to work on this code. If not, no worries, then someone else will take this PR and try to refactor it a bit. Thanks!

If something is unclear, please ask about it.

noahcabanas commented:
@markotoplak Thanks for getting back to me, I appreciate all the feedback! I'm going to try and clean up some of the issues: working with only plotted data, being more robust, simplifying infinite line, and rewriting the documentation to be direct. One part I'm not sure how to handle is redrawing when reopening a file as I used peak_state as a means to accomplish this.

markotoplak commented:
Thanks @noahcabanas! About `peak_state`: maybe I imagined what it does incorrectly, so do not worry about it too much. After your changes, I'll try to look again if I can remove it. Maybe I will not be able to. :)

noahcabanas commented:
@markotoplak I believe I've addressed most of the comments. However, for the redrawing upon open, I'm trying to use the list of peaks I made in the most recent commit called `self.peak_positions` this stored all the peaks objects so you now it will actually plot the peaks as you had them. I'm not sure where to place it to best redraw all the peaks when you reopen a canvas as there are multiple instances where the plotview is cleared. Perhaps I'm misunderstanding the application of settings?

markotoplak commented on orangecontrib/spectroscopy/widgets/owspectra.py:
```diff
@@ -1151,15 +1137,19 @@ def peak_apply_auto(self, prominence, minHeight, maxHeight, line_overlap):
 
     def delete_selected_labels(self):
-        for i, val in enumerate(self.line_list):
+        for i, val in enumerate(self.peak_positions):
```
This function does not work for clearing multiple labels: the first deleted one will be correct, all others will not because of the shift. Possible solution: go through peak_position in the reverse order.

markotoplak commented:
Thanks for the fixes. We should not store custom objects in widget settings. There would be problems if we modified that object later in any significant way: the settings then would not unpack properly.

The settings should only be lists, tuples, dicts, ints, floats, strings, bytes and None (basically whatever can be cleanly packed into json), see biolab/orange-widget-base#20

noahcabanas commented:
@markotoplak I stored the information now as a list of the position and updated span of each peak. I tried a few ways to get the data to save when a label is moved but I wasn't able to get it to work and currently have a manual save button for peak labels.

markotoplak commented:
Superseeded by #586
